aws_flow_log - adding support for sending to S3

ryandeivert opened the pull request:
<!--- Information about referencing Github Issues: https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests --->
Fixes #5482

Changes proposed in this pull request:

* Updating `aws_flow_log` resource for s3 delivery support, also improving schema validation.
* Adding state migration for `aws_flow_log` resource from v0 to v1
* Adding test for `aws_flow_log` state migration
* Updating docs for new arguments in `aws_flow_log` resource
* Updates to the 1-to-2 upgrade guide for this resource (thanks @ewbankkit)
* Way better tests from @ewbankkit


Output from acceptance testing (from @ewbankkit):

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSFlowLog_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSFlowLog_ -timeout 120m
=== RUN   TestAccAWSFlowLog_VPCID
=== PAUSE TestAccAWSFlowLog_VPCID
=== RUN   TestAccAWSFlowLog_SubnetID
--- PASS: TestAccAWSFlowLog_SubnetID (30.24s)
=== RUN   TestAccAWSFlowLog_LogDestinationType_CloudWatchLogs
--- PASS: TestAccAWSFlowLog_LogDestinationType_CloudWatchLogs (24.43s)
=== RUN   TestAccAWSFlowLog_LogDestinationType_S3
--- PASS: TestAccAWSFlowLog_LogDestinationType_S3 (28.18s)
=== CONT  TestAccAWSFlowLog_VPCID
--- PASS: TestAccAWSFlowLog_VPCID (40.73s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	124.117s
```

ewbankkit commented on aws/resource_aws_flow_log.go:
```diff
@@ -18,45 +19,71 @@ func resourceAwsFlowLog() *schema.Resource {
-				Optional:      true,
-				ForceNew:      true,
-				ConflictsWith: []string{"eni_id", "vpc_id"},
+			"resource_id": &schema.Schema{
```
Should be named `resource_ids` as potentially multiple IDs can be specified?

ryandeivert replied:
I did this to conform to the prior code but will make the change

ewbankkit commented on aws/resource_aws_flow_log.go:
```diff
@@ -65,34 +92,13 @@ func resourceAwsFlowLog() *schema.Resource {
 		LogGroupName:             aws.String(d.Get("log_group_name").(string)),
-		ResourceIds:              []*string{aws.String(resourceId)},
-		ResourceType:             aws.String(resourceType),
+		ResourceIds:              aws.StringSlice(d.Get("resource_id").([]string)),
```
As `resource_id` is declared as `schema.TypeSet`, `d.Get("resource_id")` is I think of type `*schema.Set`, so some manipulation is required to pass to `aws.StringSlice`.

ewbankkit commented on aws/resource_aws_flow_log.go:
```diff
@@ -136,17 +142,16 @@ func resourceAwsLogFlowRead(d *schema.ResourceData, meta interface{}) error {
+	d.Set("resource_id", fl.ResourceId)
 
-	var resourceKey string
 	if strings.HasPrefix(*fl.ResourceId, "vpc-") {
```
I have been thinking of various places this kind of code is used/useful (e.g. I've been thinking about solutions for https://github.com/terraform-providers/terraform-provider-aws/issues/26).
Do you think we should add some generic EC2 resource ID parsing function that returns the resource type from the ID (e.g. based on rules [here](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/resource-ids.html))?

ewbankkit commented:
Do we need to consider a diff suppression for the case

```hcl
resource "aws_flow_log" "foo" {
  log_group_name = "bar"
}
```

=>

```hcl
resource "aws_flow_log" "foo" {
  log_destination_type = "cloud-watch-logs"
  log_destination      = "bar"
}
```

and vice-versa?

ewbankkit commented on aws/resource_aws_flow_log.go:
```diff
@@ -18,45 +19,71 @@ func resourceAwsFlowLog() *schema.Resource {
+				}, false),
 			},
 
-			"vpc_id": &schema.Schema{
```
Removing existing attributes will require a deprecation warning and support of the attributes until deprecation.

ryandeivert commented:
thanks @ewbankkit I've addressed some of your suggestions. going to take a look at the diff suppression. I'm not sure how to address your suggestion of generic ec2 resource parsing but can look later as well

ryandeivert commented:
@ewbankkit I addressed the diff suppression for log destination, do you think I also have to add diff suppression to check if the old format matches the new format.. ie:

```
```

ewbankkit commented:
Yes, strictly speaking those `resource_type` diffs should be suppressed also.
The deprecations should get onto the [Terraform AWS Provider Version 2 Upgrade Guide](https://www.terraform.io/docs/providers/aws/guides/version-2-upgrade.html) also. Not sure where the source for that is.

ryandeivert commented:
good to know - I'll hunt for it. thanks for the pointers

mikecook commented:
aws/resource_aws_flow_log.go now has conflicts, could you rebase please?

tomelliff commented:
Upgrade guide is at [website/docs/guides/version-2-upgrade.html.md](https://github.com/terraform-providers/terraform-provider-aws/blob/master/website/docs/guides/version-2-upgrade.html.md).

mikecook commented:
@ryandeivert  aws/resource_aws_flow_log.go now has conflicts, could you rebase please?

ewbankkit commented:
@ryandeivert I can help with running acceptance tests when you're ready.

ryandeivert commented:
thanks @ewbankkit I've actually been running them now, trying to sort out issues. will follow up soon

uovobw commented:
greetings, this might turn out to be very useful, any way i can help? if not, is there an estimated release date? thanks all for the great work!

ryandeivert commented:
hi @uovobw I'm going to try to dedicate some time to this later today. I'm having some issues with acceptance tests (thinking related to the diff suppression) so I'll probably just commit what I have and outsource some assistance from someone here. thanks for your patience!

uovobw commented:
@ryandeivert ok, in the coming days i'll provide what help i can, thanks for the update and for your work.

AzCii commented:
exactly what i'm looking for :-) any ETA on this?

uovobw commented:
@ryandeivert sorry for the long silence, i have been _swamped_ trying to put out fires. i'll sync soon, thanks for yout effort

ewbankkit commented:
@ryandeivert I have added to your changes in https://github.com/ewbankkit/terraform-provider-aws/tree/ryandeivert-add-flow-log-s3-dest and now all acceptance tests are passing:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSFlowLog_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -run=TestAccAWSFlowLog_ -timeout 120m
=== RUN   TestAccAWSFlowLog_importBasic
--- PASS: TestAccAWSFlowLog_importBasic (28.77s)
=== RUN   TestAccAWSFlowLog_vpcToCloudWatch
--- PASS: TestAccAWSFlowLog_vpcToCloudWatch (42.33s)
=== RUN   TestAccAWSFlowLog_vpcToCloudWatchStartWithNewSyntax
--- PASS: TestAccAWSFlowLog_vpcToCloudWatchStartWithNewSyntax (42.38s)
=== RUN   TestAccAWSFlowLog_subnetToCloudWatch
--- PASS: TestAccAWSFlowLog_subnetToCloudWatch (62.23s)
=== RUN   TestAccAWSFlowLog_eniToCloudWatch
--- PASS: TestAccAWSFlowLog_eniToCloudWatch (52.06s)
=== RUN   TestAccAWSFlowLog_vpcToS3
--- PASS: TestAccAWSFlowLog_vpcToS3 (30.54s)
=== RUN   TestAccAWSFlowLog_migrateState
--- PASS: TestAccAWSFlowLog_migrateState (1.23s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	259.565s
```

You can merge back to your branch from my branch and the PR should be ready for an official review.
The trick is to mark all the deprecated attributes and the attributes that replace them as `Computed: true` . This allows a `CustomizeDiff` function to be able to modify the diff on the fly.
We canonicalize the state during state migration and on creation to store only the new attributes and then in the `CustomizeDiff` function we change the diff to the canonical form also.

ryandeivert commented:
@ewbankkit thank you!! I was suspicious of toggling the `Computed` flag, and even started to.. then got complaints about the ones I changed. I hadn't realized it was an all-or-nothing thing for a resource. I'm going to merge your changes in now - much appreciated!!

ryandeivert commented:
@ewbankkit just rebased - I think I've resolved conflicts properly but could use one more pass if you have a second. again, thank you a ton!

ewbankkit commented:
@ryandeivert I took the latest code, merged `master` to remove those conflicts and reran the acceptance tests. All are fine for me:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSFlowLog_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -run=TestAccAWSFlowLog_ -timeout 120m
=== RUN   TestAccAWSFlowLog_importBasic
--- PASS: TestAccAWSFlowLog_importBasic (29.04s)
=== RUN   TestAccAWSFlowLog_vpcToCloudWatch
--- PASS: TestAccAWSFlowLog_vpcToCloudWatch (42.49s)
=== RUN   TestAccAWSFlowLog_vpcToCloudWatchStartWithNewSyntax
--- PASS: TestAccAWSFlowLog_vpcToCloudWatchStartWithNewSyntax (42.90s)
=== RUN   TestAccAWSFlowLog_subnetToCloudWatch
--- PASS: TestAccAWSFlowLog_subnetToCloudWatch (47.44s)
=== RUN   TestAccAWSFlowLog_eniToCloudWatch
--- PASS: TestAccAWSFlowLog_eniToCloudWatch (51.95s)
=== RUN   TestAccAWSFlowLog_vpcToS3
--- PASS: TestAccAWSFlowLog_vpcToS3 (30.31s)
=== RUN   TestAccAWSFlowLog_migrateState
--- PASS: TestAccAWSFlowLog_migrateState (1.36s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	262.587s
```

The latest code is on my branch: https://github.com/ewbankkit/terraform-provider-aws/commits/ryandeivert-add-flow-log-s3-dest. Feel free to merge back.

ryandeivert commented:
thanks again @ewbankkit - pretty bizarre but I just rebased from your branch and pushed updates here to no avail. I've fetched changes and done `git diff` to ensure my branch matches yours precisely and still seeing the CI failure 🤔 despite local tests passing:
```make
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSFlowLog_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -run=TestAccAWSFlowLog_ -timeout 120m
=== RUN   TestAccAWSFlowLog_importBasic
--- PASS: TestAccAWSFlowLog_importBasic (24.71s)
=== RUN   TestAccAWSFlowLog_vpcToCloudWatch
--- PASS: TestAccAWSFlowLog_vpcToCloudWatch (35.93s)
=== RUN   TestAccAWSFlowLog_vpcToCloudWatchStartWithNewSyntax
--- PASS: TestAccAWSFlowLog_vpcToCloudWatchStartWithNewSyntax (35.30s)
=== RUN   TestAccAWSFlowLog_subnetToCloudWatch
--- PASS: TestAccAWSFlowLog_subnetToCloudWatch (39.36s)
=== RUN   TestAccAWSFlowLog_eniToCloudWatch
--- PASS: TestAccAWSFlowLog_eniToCloudWatch (44.21s)
=== RUN   TestAccAWSFlowLog_vpcToS3
--- PASS: TestAccAWSFlowLog_vpcToS3 (25.08s)
=== RUN   TestAccAWSFlowLog_migrateState
--- PASS: TestAccAWSFlowLog_migrateState (1.08s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	206.370s
```

ewbankkit commented:
Yes, I get the same when I run `make test` on the code.
Investigating...

ewbankkit commented:
@ryandeivert OK, it was my fault. I modified `TestAccAWSFlowLog_migrateState` to use some acceptance test functionality (so that we can create an ARN from the CloudWatch log group name) but it wasn't protected via the usual `resource.Test()` stuff so we need to add an explicit check for the `TF_ACC` environment variable. You can either take add the latest commit from my branch or just add

```go
	if os.Getenv(resource.TestEnvVar) == "" {
		t.Skip()
		return
	}
```

as the first code in `TestAccAWSFlowLog_migrateState`.

```console
$ make test
==> Checking that code complies with gofmt requirements...
go test ./... -timeout=30s -parallel=4
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1.838s
```

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSFlowLog_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -run=TestAccAWSFlowLog_ -timeout 120m
=== RUN   TestAccAWSFlowLog_importBasic
--- PASS: TestAccAWSFlowLog_importBasic (28.37s)
=== RUN   TestAccAWSFlowLog_vpcToCloudWatch
--- PASS: TestAccAWSFlowLog_vpcToCloudWatch (44.36s)
=== RUN   TestAccAWSFlowLog_vpcToCloudWatchStartWithNewSyntax
--- PASS: TestAccAWSFlowLog_vpcToCloudWatchStartWithNewSyntax (41.13s)
=== RUN   TestAccAWSFlowLog_subnetToCloudWatch
--- PASS: TestAccAWSFlowLog_subnetToCloudWatch (45.44s)
=== RUN   TestAccAWSFlowLog_eniToCloudWatch
--- PASS: TestAccAWSFlowLog_eniToCloudWatch (50.63s)
=== RUN   TestAccAWSFlowLog_vpcToS3
--- PASS: TestAccAWSFlowLog_vpcToS3 (36.42s)
=== RUN   TestAccAWSFlowLog_migrateState
--- PASS: TestAccAWSFlowLog_migrateState (1.29s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	247.655s
```

ryandeivert commented:
ah damn, thank you! I thought I was going crazy for a bit there. thanks for hunting it down and addressing, you've been a huge help with all this and it's much appreciated

I've rebased w/ your commit and pushed now :)

ewbankkit commented:
Success 😄!

myn6 commented:
Hello All, is there an ETA on this PR?  It is exactly what I need for me flowlogs

AzCii commented:
What is the process for getting this live?
I'm trying to understand what we are waiting for :-)

lazalo commented:
I'm interested in this as well, VPC Flow Logs through cloudwatch are very costly and would reduce the cost tremendously.

bflad commented:
This will get a first review next week. 👍

ewbankkit commented:
@bflad Agreed that the currently proposed changes are large and a more lightweight approach may be more appropriate to land this much-requested feature as soon as makes sense.
The proposed alternative changes look sensible.
I'll let @ryandeivert make the call as he's the owner of the PR.

ryandeivert commented:
@bflad / @ewbankkit I'm more than happy with a more minimal approach to this. would you suggest that I modify this PR or open a new one with those changes?

ryandeivert commented:
thanks @bflad I just reset my commits here and cherry picked your commit. output of acceptance tests:
```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSFlowLog_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSFlowLog_ -timeout 120m
=== RUN   TestAccAWSFlowLog_VPCID
=== PAUSE TestAccAWSFlowLog_VPCID
=== RUN   TestAccAWSFlowLog_SubnetID
--- PASS: TestAccAWSFlowLog_SubnetID (30.24s)
=== RUN   TestAccAWSFlowLog_LogDestinationType_CloudWatchLogs
--- PASS: TestAccAWSFlowLog_LogDestinationType_CloudWatchLogs (24.43s)
=== RUN   TestAccAWSFlowLog_LogDestinationType_S3
--- PASS: TestAccAWSFlowLog_LogDestinationType_S3 (28.18s)
=== CONT  TestAccAWSFlowLog_VPCID
--- PASS: TestAccAWSFlowLog_VPCID (40.73s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	124.117s
```

amine250 commented:
Hi, when can we expect a Terraform release that's having this change?

bflad commented:
@amine250 we typically release on Wednesdays, see also: https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md

bflad commented:
This has been released in [version 1.42.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1420-october-31-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

amine250 commented:
How can I make use of a aws_flow_log that logs on S3 inside a aws_subnet resource?
I have multiple subnets like:

```
resource "aws_subnet" "webapp" {
  availability_zone = "${data.aws_region.current.name}a"
  tags {
    Name               = "MyWebAppSubnetName"
  }
}
```

Shouldn't there be an argument for aws_subnet such as :

```
resource "aws_subnet" "webapp" {
  flow_logs {
    log_destination      = "${aws_s3_bucket.example.arn}"
    log_destination_type = "s3"
    traffic_type         = "ALL"
    vpc_id               = "${aws_vpc.example.id}"
  }
}
```

tomelliff commented:
@amine250 What makes you think it should be a configuration option on the `aws_subnet` resource instead of just being created in the `aws_flow_log` resource?

amine250 commented:
@tomelliff I'm new to Terraform so I could be saying BS.
Would you please show me a working example? Thanks.

tomelliff commented:
`flow_logs` isn't a parameter for the `aws_subnet` resource. If you want to add a flow log for a subnet then you should use the `aws_flow_log` resource and specify a `subnet_id` to point to your subnet.

If you're running into issues with that then you should raise a separate issue rather than reply to this pull request which added the S3 support for the `aws_flow_log` resource.

amine250 commented:
I think we are missing an attribute to specify S3 prefix for an aws_flow_log resource.
